### PR TITLE
Panel: added possibility to group plugins in a panel according to a grid.

### PR DESCRIPTION
**Pull request by mckn**
**What this PR does / why we need it**:
Played around with a grid layout within a panel during hack day.

**Example**:

exmaple

_Note: look a bit laggy in the gif due to low frame rate._

### CODE REVIEW

File: public/app/plugins/panel/gauge/GaugePanel.tsx [hunk 1]
```diff
@@ -61,6 +61,7 @@ export class GaugePanel extends PureComponent<PanelProps<GaugeOptions>> {
         source={data}
         renderCounter={renderCounter}
         orientation={VizOrientation.Auto}
+        itemAsSquares={true}
```
Comment (ryantxu) [inline public/app/plugins/panel/gauge/GaugePanel.tsx]: rather than `itemAsSquares` could VizOrientation have a "square" or "grid" mode?

Comment (mckn) [reply public/app/plugins/panel/gauge/GaugePanel.tsx]: Yes, I was thinking about that one as well. But wasn't sure if we wanted to enable this for some of the visualizations. So basically this is the "auto" mode for the visualizations where it makes sense to display them as squares.

But love the feedback and will change it when I continue on this one. Thanks! 🙌

**Comment by ryantxu** [comment]
@mckn -- this is looking great.  I updated it so that it fills up a whole rectangle, not just the square.  And added the option to a dropdown for the other visualizations.  I think it works well for stat panel:

image

TODO (maybe) this removed any calculation for margin -- I tried all kinds of variations but I don't think we have a simple CSS approach that will work without changing the layout or setting the margins on each item explicitly.  @torkelo  any ideas?

File: packages/grafana-ui/src/utils/squares.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  const horizontal = calculateSizeOfChild(parentHeight, parentWidth, numberOfChildren);
+  const square = Math.max(vertical, horizontal);
+  const xCount = Math.floor(parentWidth / square);
+  const yCount = Math.ceil(numberOfChildren / xCount);
+
+  return {
+    width: parentWidth / xCount,
+    height: parentHeight / yCount,
+    xCount,
+    yCount,
```
Comment (ryantxu) [inline packages/grafana-ui/src/utils/squares.ts]: don't need to return this if we give up on the commented out margin calculations above

**Comment by torkelo** [comment]
Sorry for hijacking this PR a bit, tried to add margins to each item but the margins are different for each item depending on if the items are at the edge. So had to switch to absolute positioning instead of flex + wrap. Absolute positioning is better here as it allowed for more dynamic widths of the last row so all space is used

Screen Shot 2020-03-28 at 23 08 26

Screen Shot 2020-03-28 at 23 10 29

Screen Shot 2020-03-28 at 23 11 09

**Review by ryantxu — APPROVED** [review]
@mckn -- this looks great.  Apologies for hijacking it -- its exciting to see how each of our efforts inspired the next to push it a bit further :)

**Comment by mckn** [comment]
> Sorry for hijacking this PR a bit, tried to add margins to each item but the margins are different for each item depending on if the items are at the edge. So had to switch to absolute positioning instead of flex + wrap. Absolute positioning is better here as it allowed for more dynamic widths of the last row so all space is used
> 
> Screen Shot 2020-03-28 at 23 08 26
> 
> Screen Shot 2020-03-28 at 23 10 29
> 
> Screen Shot 2020-03-28 at 23 11 09

Looks awesome!!

**Comment by mckn** [comment]
> @mckn -- this looks great. Apologies for hijacking it -- its exciting to see how each of our efforts inspired the next to push it a bit further :)

no, worries! Love the end result! Looks really good :)

**Comment by srclosson** [comment]
@mckn, @ryantxu, @torkelo Is this going to get merged? It's great :) I also loved seeing the collaboration here. Now, let's squash it!

**Comment by mckn** [comment]
> @mckn, @ryantxu, @torkelo Is this going to get merged? It's great :) I also loved seeing the collaboration here. Now, let's squash it!

Thanks, and I agree! 

There are one small thing left to do. We need to figure out how and for what visualizations the `Grid` option should be available. Me and Torkel talked a little bit about that earlier this week but we have both been busy with other things 🙈.

All ideas on how to configure this are more than welcome.

**Comment by ryantxu** [comment]
I don't think we want 'Grid' as an option (yes I know I added it!)

I think the visualization should have a property (not end user configurable) that says "supportsGrid" and then "Auto" should use grid rather than pick between horizontal/vertical.  The end behavior is the same for layouts that make sense.

**Comment by mckn** [comment]
> I don't think we want 'Grid' as an option (yes I know I added it!)
> 
> I think the visualization should have a property (not end user configurable) that says "supportsGrid" and then "Auto" should use grid rather than pick between horizontal/vertical. The end behavior is the same for layouts that make sense.

Agree, that was what I meant earlier in the thread when adding the `itemAsSquares`. So we just need to figure out a good way of controlling this and for what visualizations we want it to be the "auto".

**Comment by torkelo** [comment]
Will get this merged next week

**Pull request merged**